File: Classes/TiCoremlRealtimeRecognitionViewProxy.m
//
//  TiCoremlRealtimeRecognitionProxy.m
//  titanium-coreml
//
//  Created by Hans Knöchel on 05.07.17.
//

#if IS_IOS_11

#import "TiCoremlRealtimeRecognitionViewProxy.h"
#import "TiUtils.h"

@implementation TiCoremlRealtimeRecognitionViewProxy

#pragma mark Internal

- (TiCaptureSession *)captureSession
{
    if (_captureSession == nil) {
        _captureSession = [[TiCaptureSession alloc] initWithCompletionHandler:^(CVImageBufferRef sampleBuffer) {
            dispatch_sync(dispatch_get_main_queue(), ^{
                _currentSampleBuffer = sampleBuffer;
            });
            [self processRecognitionWithSampleBuffer:sampleBuffer];
        }];
        
        AVCaptureVideoPreviewLayer *previewLayer = [_captureSession previewLayer];
        
        [[[self view] layer] addSublayer:previewLayer];
        [self adjustFrame];
    }
    
    return _captureSession;
}

- (void)adjustFrame
{
    [_captureSession previewLayer].frame = [self view].bounds;
}

- (VNCoreMLRequest *)request
{
    if (_request == nil) {
        // Setup CoreML model
        NSError *visionModelError = nil;
        NSError *modelError = nil;
        MLModel *model = [MLModel modelWithContentsOfURL:[TiUtils toURL:[self valueForKey:@"model"] proxy:self]
                                                   error:&modelError];
        
        if (modelError != nil) {
            [self throwException:@"Error creating model" subreason:[modelError localizedDescription] location:CODELOCATION];
        }
        
        // Setup Vision CoreML model
        VNCoreMLModel *visionModel = [VNCoreMLModel modelForMLModel:model
                                                              error:&visionModelError];
        
        if (visionModelError != nil) {
            [self throwException:@"Error creating vision model" subreason:[visionModelError localizedDescription] location:CODELOCATION];
        }
        
        // Setup Vision CoreML request
        _request = [[VNCoreMLRequest alloc] initWithModel:visionModel completionHandler:^(VNRequest *request, NSError *error) {
            NSArray<VNClassificationObservation *> *observations = [request results];
            
            NSMutableArray<NSDictionary<NSString *, id> *> *result = [NSMutableArray arrayWithCapacity:[observations count]];
            
            for (VNClassificationObservation *observation in observations) {
                [result addObject:@{
                    @"identifier": observation.identifier,
                    @"confidence": NUMFLOAT(observation.confidence),
                }];
            }
            TiThreadPerformOnMainThread(^{
                [self fireEvent:@"classification" withObject:@{@"classifications": result}];
            }, NO);
        }];
        
        _request.imageCropAndScaleOption = VNImageCropAndScaleOptionCenterCrop;
    }
    
    return _request;
}

- (void)processRecognitionWithSampleBuffer:(CVImageBufferRef)sampleBuffer
{
    NSError *handlerError = nil;
    VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCVPixelBuffer:sampleBuffer options:@{}];
    
    [handler performRequests:@[[self request]] error:&handlerError];
    
    if (handlerError != nil) {
        NSLog([NSString stringWithFormat:@"[ERROR] Error processing buffer: %@", [handlerError localizedDescription]]);
    }
}

#pragma mark Public API's

- (void)startRecognition:(id)unused
{
    if (![[[self captureSession] captureSession] isRunning]) {
        [[self captureSession] start];
    } else {
        NSLog(@"[ERROR] Trying to start a capture session that is already running!");
    }
}

- (void)stopRecognition:(id)unused
{
    if ([[[self captureSession] captureSession] isRunning]) {
        [[self captureSession] stop];
    } else {
        NSLog(@"[ERROR] Trying to stop a capture session that is not running!");
    }
}

- (id)isRecognizing:(id)unused
{
    return NUMBOOL([[[self captureSession] captureSession] isRunning]);
}

- (void)takePicture:(id)value
{
    ENSURE_SINGLE_ARG(value, KrollCallback);
    
    CIImage *ciImage = [CIImage imageWithCVPixelBuffer:_currentSampleBuffer];
    CIContext *context = [CIContext contextWithOptions:nil];
    CGImageRef image = [context createCGImage:ciImage fromRect:CGRectMake(0, 0, CVPixelBufferGetWidth(_currentSampleBuffer), CVPixelBufferGetHeight(_currentSampleBuffer))];
    
    [(KrollCallback*)value call:@[@{@"image": [[TiBlob alloc] initWithImage:[UIImage imageWithCGImage:image]]}] thisObject:self];
}

@end

#endif

File: Classes/TiCoremlRealtimeRecognitionViewProxy.swift
//
//  TiCoremlRealtimeRecognitionViewProxy.swift
//  TiCoreml
//
//  Created by Hans Knöchel on 25.02.24.
//

import CoreVideo
import TitaniumKit
import UIKit
import Vision

@objc(TiCoremlRealtimeRecognitionViewProxy)
class TiCoremlRealtimeRecognitionViewProxy: TiViewProxy {

  private var currentSampleBuffer: CVImageBuffer?
  private var previewView: UIView?
  lazy private var request: VNCoreMLRequest = {
    guard let url = TiUtils.toURL(self.value(forKey: "model") as? String, proxy: self) else {
      logErrorAndFail("Cannot load model from URL")
      fatalError()
    }
    
    if url.pathExtension == "modelc" {
      logErrorAndFail("Please pass the .model file, not the .modelc")
    }

    guard let model = try? MLModel(contentsOf: MLModel.compileModel(at: url)) else {
      logErrorAndFail("Error creating model")
      fatalError()
    }

    var visionModel: VNCoreMLModel!
    
    do {
      visionModel = try VNCoreMLModel(for: model)
    } catch {
      logErrorAndFail("Error creating vision model: \(error.localizedDescription)")
      fatalError()
    }

    let _request = VNCoreMLRequest(model: visionModel) { request, error in
      guard let observations = request.results as? [VNClassificationObservation] else {
        return
      }
      
      let result: [[String: Any]] = observations.map {
        [
          "identifier": $0.identifier,
          "uuid": $0.uuid.uuidString,
          "confidence": $0.confidence
        ]
      }

      DispatchQueue.main.async {
        self.fireEvent("classification", with: ["classifications": result])
      }
    }
      
    _request.imageCropAndScaleOption = .centerCrop;
    
    return _request;
  }()
  
  private var _captureSession: TiCaptureSession!
  
  private func getCaptureSession() -> TiCaptureSession {
    guard _captureSession == nil else {
      return _captureSession
    }
    
    _captureSession = TiCaptureSession(completionHandler: { imageBuffer in
      self.currentSampleBuffer = imageBuffer
      self.processRecognitionWithSampleBuffer()
    })
    
    self.view.layer.addSublayer(_captureSession.previewLayer)
    self.adjustFrame()
    
    return _captureSession
  }
  
  func adjustFrame() {
    getCaptureSession().previewLayer.frame = view.bounds;
  }
  
  private func processRecognitionWithSampleBuffer() {
    guard let sampleBuffer = currentSampleBuffer else { return }
    let handler = VNImageRequestHandler(cvPixelBuffer: sampleBuffer)
    
    do {
      try handler.perform([request])
    } catch {
      logErrorAndFail("Error processing buffer: \(error.localizedDescription)")
    }
  }
}

// MARK: Public APIs

extension TiCoremlRealtimeRecognitionViewProxy {
  
  @objc(startRecognition:)
  func startRecognition(unused: Any?) {
    if getCaptureSession().captureSession.isRunning {
      return logErrorAndFail("Trying to start a capture session that is already running!")
    }
    
    getCaptureSession().start()
  }
  
  @objc(stopRecognition:)
  func stopRecognition(unused: Any?) {
    if !getCaptureSession().captureSession.isRunning {
      return logErrorAndFail("Trying to stop a capture session that is not running!");
    }
    
    getCaptureSession().stop()
  }
  
  @objc(isRecognizing:)
  func isRecognizing(unused: Any?) -> Bool {
    return getCaptureSession().captureSession.isRunning
  }
  
  @objc(takePicture:)
  func takePicture(value: [Any]) {
    guard let sampleBuffer = currentSampleBuffer,
          let calback = value.first as? KrollCallback else {
      return
    }
    
    let ciImage = CIImage(cvPixelBuffer: sampleBuffer)
    let context = CIContext()
    
    guard let image = context.createCGImage(ciImage, from: CGRect(x: 0, y: 0, width: CVPixelBufferGetWidth(sampleBuffer), height: CVPixelBufferGetHeight(sampleBuffer))) else {
      return
    }
    
    calback.call([["image": TiBlob(image: UIImage(cgImage: image))]], thisObject: self)
  }
}
